### PR TITLE
Fix ReferenceCounter

### PR DESCRIPTION
**Pull request by erikzhang**
Fixes neo-project/neo-devpack-dotnet#147 https://github.com/neo-project/neo/issues/1330

### CODE REVIEW

**Comment by erikzhang** [comment]
https://github.com/neo-project/neo-vm/blob/f29d46460df8507899ff955363c6f3b90a807337/src/neo-vm/ReferenceCounter.cs#L83-L88

When we are removing the zero-referred items, we only remove them from `counter`, but didn't remove them from the others items' `ObjectReferences`. (This is correct because of performance.) So when trying to remove other zero-referenced items later, we will detect these "dead" referees.

https://github.com/neo-project/neo-vm/blob/f29d46460df8507899ff955363c6f3b90a807337/src/neo-vm/ReferenceCounter.cs#L76-L78

So we can just ignore the "dead" referees because they have already been removed.

**Comment by erikzhang** [comment]
https://github.com/neo-project/neo-vm/blob/f29d46460df8507899ff955363c6f3b90a807337/src/neo-vm/ReferenceCounter.cs#L68

If the `entry` doesn't exist, we can just ignore it because it has already been removed.

File: src/neo-vm/ReferenceCounter.cs [hunk 1]
```diff
@@ -65,7 +65,7 @@ internal int CheckZeroReferred()
                     while (toBeChecked.Count > 0)
                     {
                         CompoundType c = toBeChecked.Dequeue();
-                        Entry entry = counter[c];
+                        if (!counter.TryGetValue(c, out Entry entry)) continue;
                         if (entry.StackReferences > 0)
                         {
                             toBeDestroyedInLoop.Clear();
```
Comment (shargon) [inline src/neo-vm/ReferenceCounter.cs]: And it should not be the same as this, and clear `toBeDestroyedInLoop`?

Comment (shargon) [reply src/neo-vm/ReferenceCounter.cs]: Ok, no because it has no references...

**Comment by erikzhang** [comment]
Please wait for my UT.

**Comment by erikzhang** [comment]
@shargon The real fix has come. 😂

**Comment by erikzhang** [comment]
You can see, withou the changes, the UT fails.

**Comment by erikzhang** [comment]
Completely fixed!

File: src/neo-vm/Types/Map.cs [hunk 2]
```diff
@@ -30,8 +31,9 @@ public class Map : CompoundType, IReadOnlyDictionary<PrimitiveType, StackItem>
         }
 
         public override int Count => dictionary.Count;
-        public override int ItemsCount => dictionary.Count * 2;
         public IEnumerable<PrimitiveType> Keys => dictionary.Keys;
+        internal override IEnumerable<StackItem> SubItems => Keys.Concat(Values);
```
Comment (shargon) [inline src/neo-vm/Types/Map.cs]: This is the final fix! :) thanks for it :D .... but if you want to make me happy. I don't like to see `Linq` inside the VM hahah.

```suggestion
        internal override IEnumerable<StackItem> SubItems()
        {
                foreach (var key in Keys) yield return key;
                foreach (var value in Values) yield return value;
        }
```

If you don't like it, I will try to convince you in the future :P

File: src/neo-vm/ReferenceCounter.cs [hunk 3]
```diff
@@ -85,7 +86,7 @@ internal int CheckZeroReferred()
                 {
                     counter.Remove(compound);
                     references_count -= compound.SubItemsCount;
-                    foreach (CompoundType subitem in compound.SubItems)
+                    foreach (CompoundType subitem in compound.SubItems.OfType<CompoundType>())
```
Comment (vncoelho) [inline src/neo-vm/ReferenceCounter.cs]: aueaheuaea, things that are hard to see.

File: src/neo-vm/Types/Map.cs (hunk 2 above)
Comment (erikzhang) [reply src/neo-vm/Types/Map.cs]: In fact they are exactly the same.

Comment (erikzhang) [reply src/neo-vm/Types/Map.cs]: This is how linq is implemented internally.

**Pull request merged**